rfcs/0022-minimal-module-list.md

```diff
@@ -0,0 +1,143 @@
+---
+feature: minimal-module-list
+start-date: 2018-01-12
+author: Eelco Dolstra
+co-authors: (find a buddy later to help our with the RFC)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Evaluating a NixOS configuration is getting ever slower due to the
+reliance on a global list of modules unconditionally included by every
+configuration. The proposal is to switch from the use of `enable`
+options to explicit inclusion of needed modules.
+
+# Motivation
+[motivation]: #motivation
+
+NixOS evaluation is getting slower with every release, as can be seen
+here:
+
+* https://hydra.nixos.org/job/nixpkgs/trunk/metrics/metric/nixos.smallContainer.time
+* https://hydra.nixos.org/job/nixpkgs/trunk/metrics/metric/nixos.lapp.time
+* https://hydra.nixos.org/job/nixpkgs/trunk/metrics/metric/nixos.kde.time
+
+Evaluator memory use grows similarly, e.g.
+
+* https://hydra.nixos.org/job/nixpkgs/trunk/metrics/metric/nixos.smallContainer.allocations
+
+This increase in resource consumption is particularly problematic for
+NixOps networks consisting of many machines.
+
+The main reason for the increase is the growth in the number of NixOS
+modules, from 497 in `module-list.nix` in NixOS 15.09 to 739 in
+17.09. The module system needs to evaluate every module, even though
+most modules are not needed for a particular configuration.
+
+The proposal is to require most modules to be included explicitly in a
+system configuration. That is, rather than writing e.g.
+
+    hardware.pulseaudio.enable = true;
+
+you write
+
+    imports = [ <nixpkgs/nixos/modules/config/pulseaudio.nix> ];
+
+For example, here is the time and memory consumption for evaluating a
+50-machine NixOps network:
+
+| Nixpkgs                     | Time (s) | RSS (MiB) |
+| --------------------------- | -------- | --------- |
+| 17.09                       | 73.7     | 6470      |
+| 17.09 minimal               | 19.5     | 2577      |
+| 17.09 minimal + memoisation | 12.5     | 1532      |
```

Comment:
12 seconds is actually quite reasonable compared to building pretty much anything else. Is it not possible to prepare the system outputs through Hydra?